api/agent/agent.go

```diff
@@ -278,55 +284,95 @@ func (a *agent) getSlot(ctx context.Context, call *call) (Slot, error) {
 	return a.launchCold(ctx, call)
 }
 
-// launchHot checks with slot queue to see if a new container needs to be launched and waits
-// for available slots in the queue for hot request execution.
-func (a *agent) launchHot(ctx context.Context, call *call) (Slot, error) {
+// hotLauncher is spawned in a go routine for each slot queue to monitor stats and launch hot
+// containers if needed. Upon shutdown or activity timeout, hotLauncher exits and during exit,
+// it destroys the slot queue.
+func (a *agent) hotLauncher(ctx context.Context, call *call) {
+
+	// Let use 60 minutes or 2 * IdleTimeout as hot queue idle timeout, pick
+	// whichever is longer. If in this time, there's no activity, then
+	// we destroy the hot queue.
+	timeout := time.Duration(60) * time.Minute
+	idleTimeout := time.Duration(call.IdleTimeout) * time.Second * 2
+	if timeout < idleTimeout {
+		timeout = idleTimeout
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
```

Comment:
better use context-bound logger https://github.com/fnproject/fn/blob/0b594cfeff7576fc685fad88823ca552766223e8/api/common/ctx.go#L17

Reply:
I could, but this is similar to runHot() where we do not use the call context (which will expire with timeout) except for spans. Should I do this? As this ctx will expire shortly, while the launcher will continue running for extended period. So, we'd be running the logger under a ctx that is expired. The code here is a bit confusing because it still uses 'ctx' and passes that to runHot(), which only uses it to start a span. I'm open to ideas here.

Reply:
runHot() for instance uses a new logrus instead of using ctx of the original call.

Reply:
we love spans, we love tracing, so why not?

Reply:
@denismakogon you mean like this below, right?

```
	logger := common.Logger(ctx).WithFields(logrus.Fields{
		"app":              call.AppName,
		"route":            call.Path,
		"image":            call.Image,
		"memory":           call.Memory,
		"format":           call.Format,
		"idle_timeout":     call.IdleTimeout,
		"launcher_timeout": timeout})
```

Reply:
Yeah, it still looks ugly, but who really cares about the beauty (maybe one more logger method to hide all of this and build logger from ctx and call?).

Reply:
Let me see what I can do here. There's some common log patterns between here and in runHot(), I'll hoist them into a common helper function.